### PR TITLE
[tests-only][full-ci] refactoring user addition to group to graphapi

### PR DESCRIPTION
**Pull request by S-Panta**
## Description
This PR refactors the user addition to group test code from ocs api to graphapi

## Related Issue

- Part of https://github.com/owncloud/ocis/issues/6998

## How Has This Been Tested?

- test environment:
- locally
- ci

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [x] Tests only (no source changes)

## Checklist:

- [ ] Code changes
- [ ] Unit tests added
- [x] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 1]
```diff
@@ -772,7 +772,7 @@ public function theUserCreatesNewUser(string $user, TableNode $table): void {
 	 */
 	public function addUserToGroup(string $group, string $user, ?string $byUser = null): ResponseInterface {
 		$credentials = $this->getAdminOrUserCredentials($byUser);
-		$groupId = $this->featureContext->getAttributeOfCreatedGroup($group, "id");
+		$groupId = $this->featureContext->getAttributeOfCreatedGroup($group, "id") ?: WebDavHelper::generateUUIDv4();
```
Comment (SwikritiT) [inline tests/acceptance/features/bootstrap/GraphContext.php:775]: I don't think we should generate a fake groupId in case the group doesn't exist in this main function. This generation of fake id is to check for the edge cases of `non-existent` group but in the case of some error the group was not created in the previous `when` or `given` step this will generate a fake id, the test will obviously fail but I think this will make the debugging very confusing? Maybe we can add a filter in the function `$createNonExistentGroup` which defaults to false and only when this filter is true generate the fake id otherwise just follow the normal flow
```suggestion
		$groupId = $this->featureContext->getAttributeOfCreatedGroup($group, "id")
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 2]
```diff
@@ -772,7 +772,7 @@ public function theUserCreatesNewUser(string $user, TableNode $table): void {
 	 */
 	public function addUserToGroup(string $group, string $user, ?string $byUser = null): ResponseInterface {
```
Comment (SwikritiT) [inline tests/acceptance/features/bootstrap/GraphContext.php:773]:
```suggestion
	public function addUserToGroup(string $group, string $user, ?string $byUser = null, bool $createNonExistentGroup = false): ResponseInterface {
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 3]
```diff
@@ -847,19 +847,7 @@ public function theAdministratorTriesToAddUserToGroupUsingTheGraphAPI(string $us
...
-			GraphHelper::addUserToGroup(
-				$this->featureContext->getBaseUrl(),
-				$this->featureContext->getStepLineRef(),
-				$credentials['username'],
-				$credentials['password'],
-				$userId,
-				$groupId
-			)
-		);
+		$this->featureContext->setResponse($this->addUserToGroup('', $user, $byUser));
```
Comment (SwikritiT) [inline tests/acceptance/features/bootstrap/GraphContext.php:850]: Would this solve what you're trying to achieve here?
```suggestion
		$this->featureContext->setResponse($this->addUserToGroup('', $user, $byUser,true));
```

File: tests/acceptance/features/bootstrap/GraphContext.php (hunk 1 above)
Comment (S-Panta) [reply tests/acceptance/features/bootstrap/GraphContext.php:775]: Since the user addition to the group always has group creation in its `given` statement (not in the `when` step), whether or not a group has been created will be checked inside that test code. The group creation using the `when` step doesn't call this function at all.

Comment (SwikritiT) [reply tests/acceptance/features/bootstrap/GraphContext.php:775]: doesn't matter having adding the fake ID when there is no group is not good logic please add a filter and set it only when we expect the scenario to have no group. Also, why is not the group creation when step not calling this function? how many `addUserToGroup` do we have? we should not have multiple function to do the same task

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 4]
```diff
@@ -824,7 +833,7 @@ public function theAdministratorAddsTheFollowingUsersToTheFollowingGroupsUsingTh
 	}
 
 	/**
-	 * @When the administrator tries to add user :user to group :group using the Graph API
+	 * @When the administrator tries to add nonexistent user :user to group :group using the Graph API
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:836]:
```suggestion
	 * @When the administrator tries to add nonexistent user to group :group using the Graph API
```

Comment (saw-jan) [reply tests/acceptance/features/bootstrap/GraphContext.php:836]: and simply pass `nonexistent` to `addUserToGroup`: 
```php
$this->featureContext->setResponse($this->addUserToGroup($group, "nonexistent"));
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 5]
```diff
@@ -847,19 +856,7 @@ public function theAdministratorTriesToAddUserToGroupUsingTheGraphAPI(string $us
...
-			GraphHelper::addUserToGroup(
-				$this->featureContext->getBaseUrl(),
-				$this->featureContext->getStepLineRef(),
-				$credentials['username'],
-				$credentials['password'],
-				$userId,
-				$groupId
-			)
-		);
+		$this->featureContext->setResponse($this->addUserToGroup('', $user, $byUser, true));
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:859]:
```suggestion
		$this->featureContext->setResponse($this->addUserToGroup('nonexistent', $user, $byUser, true));
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 6]
```diff
@@ -766,14 +766,23 @@ public function theUserCreatesNewUser(string $user, TableNode $table): void {
 	 * @param string $group
 	 * @param string $user
 	 * @param string|null $byUser
+	 * @param bool $createNonExistentGroup
 	 *
 	 * @return ResponseInterface
 	 * @throws GuzzleException
 	 */
-	public function addUserToGroup(string $group, string $user, ?string $byUser = null): ResponseInterface {
+	public function addUserToGroup(string $group, string $user, ?string $byUser = null, bool $createNonExistentGroup = false): ResponseInterface {
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:774]:
```suggestion
	public function addUserToGroup(string $group, string $user, ?string $byUser = null): ResponseInterface {
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 7]
```diff
@@ -766,14 +766,23 @@ public function theUserCreatesNewUser(string $user, TableNode $table): void {
...
+		if ($createNonExistentGroup) {
+			$groupId = WebDavHelper::generateUUIDv4();
+		} else {
+			$groupId = $this->featureContext->getAttributeOfCreatedGroup($group, "id");
+		}
+		if ($this->featureContext->userExists($user)) {
+			$userId = $this->featureContext->getAttributeOfCreatedUser($user, "id");
+		} else {
+			$userId = WebDavHelper::generateUUIDv4();
+		}
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:785]:
```suggestion
		if ($group === "nonexistent") {
			$groupId = WebDavHelper::generateUUIDv4();
		} else {
			$groupId = $this->featureContext->getAttributeOfCreatedGroup($group, "id");
		}
		if ($user === "nonexistent") {
			$userId = WebDavHelper::generateUUIDv4();
		} else {
			$userId = $this->featureContext->getAttributeOfCreatedUser($user, "id");
		}
```

**Review by saw-jan — COMMENTED** [review]
IMO, using `nonexistent` constantly should do the trick and we don't have to add extra param

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 8]
```diff
@@ -762,6 +762,8 @@ public function theUserCreatesNewUser(string $user, TableNode $table): void {
 
 	/**
 	 * adds a user to a group
+	 * Note that the string "nonexistent" shouldn't be used for username and groupname
+	 * Only for nonexistent testcase
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:766]:
```suggestion
	 * NOTE: If you want to do request with non-existing user or group then provide "nonexistent" as the name
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 9]
```diff
@@ -772,8 +774,18 @@ public function theUserCreatesNewUser(string $user, TableNode $table): void {
 	 */
 	public function addUserToGroup(string $group, string $user, ?string $byUser = null): ResponseInterface {
 		$credentials = $this->getAdminOrUserCredentials($byUser);
-		$groupId = $this->featureContext->getAttributeOfCreatedGroup($group, "id");
-		$userId = $this->featureContext->getAttributeOfCreatedUser($user, "id");
+		//user shouldn't be created with "nonexistent" name but only used when for nonexistent group testcase
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:777]:
```suggestion
		// NOTE: If you want to do request with non-existing user or group then provide "nonexistent" as the name
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 10]
```diff
@@ -772,8 +774,18 @@ public function theUserCreatesNewUser(string $user, TableNode $table): void {
...
+		if ($group === "nonexistent") {
+			$groupId = WebDavHelper::generateUUIDv4();
+		} else {
+			$groupId = $this->featureContext->getAttributeOfCreatedGroup($group, "id");
+		}
+		//user shouldn't be created with "nonexistent" name but used when for nonexistent users testcase
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:783]:
```suggestion
		// NOTE: If you want to do request with non-existing user or group then provide "nonexistent" as the name
```

File: tests/acceptance/features/bootstrap/Provisioning.php [hunk 11]
```diff
@@ -2966,23 +2966,7 @@ public function userBelongsToGroup(string $user, string $group):bool {
...
-	 * @return void
-	 * @throws Exception
-	 */
-	public function theAdministratorAddsUserToTheFollowingGroupsUsingTheProvisioningApi(TableNode $table):void {
-		$this->verifyTableNodeColumns($table, ["username", "groupname"], ["comment"]);
-		$rows = $table->getHash();
-		foreach ($rows as $row) {
-			$this->adminAddsUserToGroupUsingTheProvisioningApi($row["username"], $row["groupname"]);
-		}
+		$this->addUserToGroup($user, $group);
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/Provisioning.php:2969]: set the response in When steps

**Review by saw-jan — APPROVED** [review]
LGTM

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 12]
```diff
@@ -762,6 +762,7 @@ public function theUserCreatesNewUser(string $user, TableNode $table): void {
 
 	/**
 	 * adds a user to a group
+	 * NOTE: If you want to do request with non-existing user or group,use provide "nonexistent" as their name
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:765]:
```suggestion
	 * NOTE: If you want to make a request with a non-existing user or group, provide "nonexistent" as their name
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 13]
```diff
@@ -772,8 +773,18 @@ public function theUserCreatesNewUser(string $user, TableNode $table): void {
 	 */
 	public function addUserToGroup(string $group, string $user, ?string $byUser = null): ResponseInterface {
 		$credentials = $this->getAdminOrUserCredentials($byUser);
-		$groupId = $this->featureContext->getAttributeOfCreatedGroup($group, "id");
-		$userId = $this->featureContext->getAttributeOfCreatedUser($user, "id");
+		// NOTE: If you want to do request with non-existing group,use "nonexistent" as the name
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:776]:
```suggestion
		// NOTE: If you want to make a request with a non-existing group, use "nonexistent" as the name
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 14]
```diff
@@ -772,8 +773,18 @@ public function theUserCreatesNewUser(string $user, TableNode $table): void {
...
+		if ($group === "nonexistent") {
+			$groupId = WebDavHelper::generateUUIDv4();
+		} else {
+			$groupId = $this->featureContext->getAttributeOfCreatedGroup($group, "id");
+		}
+		// NOTE: If you want to do request with non-existing user,use "nonexistent" as the name
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:782]:
```suggestion
		// NOTE: If you want to make a request with a non-existing user, use "nonexistent" as the name
```

File: tests/acceptance/features/bootstrap/Provisioning.php [hunk 15]
```diff
@@ -3147,10 +3046,10 @@ public function addUserToGroup(string $user, string $group, ?string $method = nu
...
 				break;
 			case "graph":
-				$this->graphContext->adminHasAddedUserToGroupUsingTheGraphApi(
-					$user,
-					$group
-				);
+				$result = $this->graphContext->addUserToGroup($group, $user);
+				if ($result->getStatusCode() !== 204) {
+					$this->throwHttpException($result, "Could not add user '$user' to group '$group'.");
+				}
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/Provisioning.php:3052]: We shouldn't use assertion within a helper function

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 16]
```diff
@@ -1987,12 +1985,8 @@ public function theAdministratorHasAddedTheFollowingUsersToAGroupAtOnceUsingTheG
 		foreach ($table->getHash() as $row) {
 			$userIds[] = $this->featureContext->getAttributeOfCreatedUser($row['username'], "id");
 		}
-		$this->addMultipleUsersToGroup($user, $userIds, $groupId, $table);
-		$response = $this->featureContext->getResponse();
-		if ($response->getStatusCode() !== 204) {
-			$this->throwHttpException($response, "Cannot add users to group '$group'");
-		}
-		$this->featureContext->emptyLastHTTPStatusCodesArray();
+		$response = $this->addMultipleUsersToGroup($user, $userIds, $groupId, $table);
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/GraphContext.php:1988]: I didn't see use for $table in `addMultipleUsersToGroup`. column verification can/should be done within stef def (wherever possible)
```suggestion
		$response = $this->addMultipleUsersToGroup($user, $userIds, $groupId);
```

File: tests/acceptance/features/bootstrap/Provisioning.php [hunk 17]
```diff
@@ -3055,7 +2981,21 @@ public function theAdministratorTriesToAddUserToGroupUsingTheProvisioningApi(
 	 */
 	public function userHasBeenAddedToGroup(string $user, string $group):void {
 		$user = $this->getActualUsername($user);
-		$this->addUserToGroup($user, $group, null, true);
+		if ($this->isTestingWithLdap() && !$this->isLocalAdminGroup($group)) {
```
Comment (saw-jan) [inline tests/acceptance/features/bootstrap/Provisioning.php:2984]:
```suggestion
		if ($this->isTestingWithLdap()) {
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 18]
```diff
@@ -772,8 +773,18 @@ public function theUserCreatesNewUser(string $user, TableNode $table): void {
 	 */
 	public function addUserToGroup(string $group, string $user, ?string $byUser = null): ResponseInterface {
 		$credentials = $this->getAdminOrUserCredentials($byUser);
-		$groupId = $this->featureContext->getAttributeOfCreatedGroup($group, "id");
-		$userId = $this->featureContext->getAttributeOfCreatedUser($user, "id");
+		// NOTE: If you want to make a request with non-existing group,provide "nonexistent" as the name
```
Comment (SagarGi) [inline tests/acceptance/features/bootstrap/GraphContext.php:776]:
```suggestion
```

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 19]
```diff
@@ -772,8 +773,18 @@ public function theUserCreatesNewUser(string $user, TableNode $table): void {
...
+		if ($group === "nonexistent") {
+			$groupId = WebDavHelper::generateUUIDv4();
+		} else {
+			$groupId = $this->featureContext->getAttributeOfCreatedGroup($group, "id");
+		}
+		// NOTE: If you want to make a request with non-existing user,provide "nonexistent" as the name
```
Comment (SagarGi) [inline tests/acceptance/features/bootstrap/GraphContext.php:782]:
```suggestion
```
I think comment in the function is enough.

**Review by saw-jan — APPROVED** [review]
LGTM :+1:

File: tests/acceptance/features/bootstrap/GraphContext.php [hunk 20]
```diff
@@ -772,8 +773,16 @@ public function theUserCreatesNewUser(string $user, TableNode $table): void {
...
+		if ($group === "nonexistent") {
+			$groupId = WebDavHelper::generateUUIDv4();
+		} else {
+			$groupId = $this->featureContext->getAttributeOfCreatedGroup($group, "id");
+		}
+		if ($user === "nonexistent") {
+			$userId = WebDavHelper::generateUUIDv4();
+		} else {
+			$userId = $this->featureContext->getAttributeOfCreatedUser($user, "id");
+		}
```
Comment (SagarGi) [inline tests/acceptance/features/bootstrap/GraphContext.php:785]:
```suggestion
		$groupId = ($group === "nonexistent") ? WebDavHelper::generateUUIDv4() : $this->featureContext->getAttributeOfCreatedGroup($group, "id");
		$$userId = ($user === "nonexistent") ? WebDavHelper::generateUUIDv4() : $userId = $this->featureContext->getAttributeOfCreatedUser($user, "id");
```
can we make short like this?

**Review by SagarGi — APPROVED** [review]
just small change Other LGTM :+1:

File: tests/acceptance/features/bootstrap/GraphContext.php (hunk 20 above)
Comment (saw-jan) [reply tests/acceptance/features/bootstrap/GraphContext.php:785]: I would prefer if-else block in this case because it's easier to read. But I am fine with both
